### PR TITLE
Lifecycle validators

### PR DESCRIPTION
**Pull request by elijahbenizzy**
Adds a lifecycle validation capability. We had two types of lifecycle customizations:
1. Methods (replaced some component)
2. Hooks (augmented some components)

Now we have validators, which return a [bool, str] result, consisting of [success, message]. These allow erroring out if conditions are not met. We add one for validating nodes, and one for validating DAGs.

We also do a bunch of repo hygiene/add a few more public-facing capabilities.

https://github.com/DAGWorks-Inc/hamilton/issues/631

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: hamilton/base.py [hunk 1]
```diff
@@ -11,15 +11,17 @@
 import pandas as pd
 from pandas.core.indexes import extension as pd_extension
 
+from hamilton.lifecycle import api as lifecycle_api
+
 try:
-    from . import htypes, lifecycle, node
+    from . import htypes, node
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply hamilton/base.py:17]: some recollection it was about testing, or exercising modules as `__main__` ?

File: hamilton/lifecycle/base.py [hunk 2]
```diff
@@ -107,13 +145,38 @@ def validate_method_fn(fn: Callable):
     validate_lifecycle_adapter_function(fn, returns_value=True)
 
 
+def validate_validator_fn(fn: Callable):
+    """Ensures that a function forms a registerable "validator". These are currently the same rules as "hooks".
+    While they should also raise an exception, that is not possible to express in the type annotation.
+
+    @param fn: Function to validate
```
Comment (skrawcz) [inline hamilton/lifecycle/base.py]:
```suggestion
    :param fn: Function to validate
```

File: hamilton/lifecycle/base.py [hunk 3]
```diff
@@ -107,13 +145,38 @@ def validate_method_fn(fn: Callable):
...
     """Container class for decorators to register hooks/methods."""
 
     @classmethod
     def base_hook(cls, fn_name: str):
         """Hooks get called at distinct stages of Hamilton's execution.
-        These can be layered together, and potentially coupled to other hooks."""
+        These can be layered together, and potentially coupled to other hooks.
+
+        @param fn_name:
+        @return:
```
Comment (skrawcz) [inline hamilton/lifecycle/base.py]:
```suggestion
        :param fn_name:
        :return:
```

File: hamilton/lifecycle/base.py [hunk 4]
```diff
@@ -161,6 +227,32 @@ def decorator(clazz):
...
         return decorator
 
+    @classmethod
+    def base_validator(cls, fn_name: str):
+        """Validators have a similar conceptual relationship to hooks. They provide custom validation
+        logic, and multiple can be layered together. That said, they *also* expect you to raise an error
+        if there is an issue, and have no output.
+
+        @param fn_name: Name of the function in the class we're registering.
+        @param raises:  Type of the exception that the validator raises.
```
Comment (skrawcz) [inline hamilton/lifecycle/base.py]:
```suggestion
        :param fn_name: Name of the function in the class we're registering.
        :param raises:  Type of the exception that the validator raises.
        :return: Callable function
```

File: hamilton/lifecycle/base.py [hunk 5]
```diff
@@ -196,27 +288,39 @@ def do_validate_input(self, *, node_type: type, input_value: Any) -> bool:
...
+    @abc.abstractmethod
+    def validate_graph(
+        self, *, graph: "graph.FunctionGraph", modules: List[ModuleType], config: Dict[str, Any]
+    ) -> Tuple[bool, Optional[Exception]]:
+        """Validates the graph. This will raise an InvalidNodeException
+
+        @param graph:
+        @param modules:
+        @param config:
+        @return:
```
Comment (skrawcz) [inline hamilton/lifecycle/base.py]:
```suggestion
        :param graph:
        :param modules:
        :param config:
        :return:
```

**Review by skrawcz — APPROVED** [review]

File: hamilton/driver.py [hunk 6]
```diff
@@ -363,6 +399,15 @@ def __init__(
...
             self.capture_constructor_telemetry(error, modules, config, adapter)
 
+    def get_graph(self) -> graph_types.HamiltonGraph:
+        """Gives a graph object that you can use -- export, inspect, walk, etc...
+        This is in place of accessing the internal graph directly, which does not commit
+        to a stable API.
+
+        :return: A HamiltonGraph object.
+        """
+        return graph_types.HamiltonGraph.from_graph(self.graph)
```
Comment (skrawcz) [inline hamilton/driver.py]: I don't think we shouldn't expose this quite so directly -- any export functions would utilize this logic though.

File: hamilton/graph_types.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+@dataclass
+class HamiltonNode:
+    """External facing API for hamilton Nodes. Having this as a dataclass allows us
+    to hide the internals of the system but expose what the user might need.
+    Furthermore, we can always add attributes and maintain backwards compatibility."""
+
+    name: str
+    type: typing.Type
+    tags: typing.Dict[str, str]
```
Comment (skrawcz) [inline hamilton/graph_types.py:24]: this
```suggestion
    tags: typing.Dict[str, Union[str, List[str]]]
```

File: hamilton/driver.py [hunk 8]
```diff
@@ -314,6 +289,66 @@ def normalize_adapter_input(
...
+                "validate_graph",
+                output_only_failures=True,
+                graph=graph,
+                modules=graph_modules,
+                config=graph.config,
+            )
+            if validation_results:
+                error_delimiter = "\t"
+                errors = [result.error for result in validation_results]
+                errors.sort()
```
Comment (elijahbenizzy) [inline hamilton/driver.py:345]: Combine into one rather than using in-place `.sort()`

File: hamilton/lifecycle/api.py [hunk 9]
```diff
@@ -227,7 +254,96 @@ def post_node_execute(
...
+    def pre_graph_execute(
+        self,
+        *,
+        run_id: str,
+        graph: "FunctionGraph",
+        final_vars: List[str],
+        inputs: Dict[str, Any],
+        overrides: Dict[str, Any],
+    ):
+        all_nodes, user_defined_nodes = graph.get_upstream_nodes(final_vars, inputs, overrides)
```
Comment (elijahbenizzy) [inline hamilton/lifecycle/api.py:286]: Add comments/docstring on all these functions -- nice to have for internal documentation, even though it's pretty clear

File: hamilton/lifecycle/api.py [hunk 10]
```diff
@@ -307,3 +427,46 @@ def run_to_execute_node(
         :return: The result of the node execution -- up to you to return this.
         """
         pass
+
+
+class StaticValidator(BaseValidateGraph, BaseValidateNode):
+    """Performs static validation of the DAG. Note that this has the option to perform default validat"""
```
Comment (elijahbenizzy) [inline hamilton/lifecycle/api.py:433]: Finish sentence

File: hamilton/lifecycle/base.py [hunk 11]
```diff
@@ -27,32 +27,59 @@
...
+# lifecycle_base has elements that use type-hinting with the FunctionGraph object
+# This is OK -- in a *real* compiled language this wouldn't be an issue
+# (you'd have header types in C, and java should be smart enough). Given that we're using
+# python, which (our usage of) leans type-hinting trigger-happy, this will suffice.
 if TYPE_CHECKING:
-    from hamilton.graph import FunctionGraph
+    from hamilton import graph, node
 
-from hamilton.node import Node
+# All of this is internal APIs. We only have a few types of lifecycle steps, and they store
```
Comment (elijahbenizzy) [inline hamilton/lifecycle/base.py:46]: Clarify this comment, it's very unclear what I mean by "individually"

File: hamilton/lifecycle/base.py [hunk 12]
```diff
@@ -61,21 +88,32 @@ class InvalidLifecycleAdapter(Exception):
     pass
 
 
-def validate_lifecycle_adapter_function(fn: Callable, returns_value: bool):
+def validate_lifecycle_adapter_function(
```
Comment (elijahbenizzy) [inline hamilton/lifecycle/base.py:91]: Add unit tests for this -- we have a new parameter

File: hamilton/lifecycle/base.py [hunk 13]
```diff
@@ -107,13 +145,37 @@ def validate_method_fn(fn: Callable):
...
+    """
+    if inspect.iscoroutinefunction(fn):
+        raise InvalidLifecycleAdapter(
+            f"Lifecycle validators must (so far) be synchronous, "
+            f"but {fn} is an async function. "
+        )
+    validate_lifecycle_adapter_function(fn, returns_value=True)
+
+
+# Container for fns to make registering hooks/methods easy -- this is just in the superclass,
```
Comment (elijahbenizzy) [inline hamilton/lifecycle/base.py:163]: Clarify comment -- this is hard to follow

File: hamilton/lifecycle/base.py [hunk 14]
```diff
@@ -161,6 +226,32 @@ def decorator(clazz):
 
         return decorator
 
+    @classmethod
+    def base_validator(cls, fn_name: str):
+        """Validators have a similar conceptual relationship to hooks. They provide custom validation
```
Comment (elijahbenizzy) [inline hamilton/lifecycle/base.py:231]: Fix up text -- make it clearer.

File: hamilton/lifecycle/base.py [hunk 15]
```diff
@@ -711,6 +837,23 @@ def does_method(self, method_name: str, is_async: bool) -> bool:
...
 
+    def does_validation(self, validator_name: str) -> bool:
+        """Whether a validator is implemented by any of the adapters in this group.
+
+        :param validator_name: Name of the validator
+        :param is_async: Whether you want the async version or not
+        :return: True if this adapter set does this validator, False otherwise
+        """
+        if validator_name not in REGISTERED_SYNC_VALIDATORS:
+            import pdb
```
Comment (elijahbenizzy) [inline hamilton/lifecycle/base.py:848]: Womp womp
